### PR TITLE
Ruby capstone, Catalog of my things

### PR DESCRIPTION
**Pull request by smunoz1988**
Hello dear reviewer! 

team: @Booze33 / @smunoz1988 / @Metaverse-Nabeel 

For this milestone, we make these changes:

Project requirements

Logic

-Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) was created in a separate .rb file.
-All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes were implemented.
-All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
-All methods visible in the diagram were implemented:
-add_item method (in all classes that have that method)

- should take an instance of the Item class as an input.
- should add the input item to the collection of items.
- should add self as a property of the item object (by using the correct setter from the item object).

-can_be_archived?() in the Item class
- should return true if published_date is older than 10 years.
- otherwise, it should return false.
-move_to_archive() in the Item class
- should reuse can_be_archived?() method.
- should change the archived property to true if the result of the can_be_archived?() method is true.
- should do nothing if the result of the can_be_archived?() method is false.
-can_be_archived?() in the Book class
- should override the method from the parent class.
- should return true if parent's method returns true OR if cover_state equals to "bad".
- otherwise, it should return false.
-can_be_archived?() in the MusicAlbum class
- should override the method from the parent class.
- should return true if parent's method returns true AND if on_spotify equals true.
- otherwise, it should return false.
-can_be_archived?() in the Movie class
- should override the method from the parent class.
- should return true if parent's method returns true OR if silent equals true.
- otherwise, it should return false.
-can_be_archived?() in the Game class
- should override the method from the parent class.
- should return true if parent's method returns true AND if last_played_at is older than 2 years.
- otherwise, it should return false.
-Add unit tests for all implemented methods.

User interface

- Create a main.rb file that will serve as your console app entry-point.
- The console app, at the start:
- Present the user with a list of options to perform.
- Let users choose an option.
- If needed, ask for parameters for the option.
- Have a way to quit the app.
- The following options should be available:
- List all books
- List all music albums
- List all movies
- List of games
- List all genres (e.g 'Comedy', 'Thriller')
- List all labels (e.g. 'Gift', 'New')
- List all authors (e.g. 'Stephen King')
- List all sources (e.g. 'From a friend', 'Online shop')
- Add a book
- Add a music album
- Add a movie
- Add a game
- All data is preserved by saving collections in .json files.

Database

We create a schema.sql file with tables that will be analogical to the structure of the classes in your app. 

image

Video link: 

https://drive.google.com/file/d/1z3AHJHSdi9sWAMS1ARBFreGRtBK73hQ5/view?usp=sharing

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,167 @@
...
+    - [Install](#install)
+    - [Usage](#usage)
+    - [Run tests](#run-tests)
+    - [Deployment](#deployment)
+  - [👥 Authors ](#-authors-)
+  - [🔭 Future Features ](#-future-features-)
+  - [🤝 Contributing ](#-contributing-)
+  - [📝 License ](#-license-)
+
+# 📖 TDD Project <a name="about-project">TDD Project</a>
```
Comment (muneebulrehman) [inline README.md]: Kindly give a proper title to your project. I believe this is the title of your last project. Remember, it is super important to have a professional and clean README file.

File: item.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def move_to_archive
+    if can_be_archived? == true
+      @archived = true
+    else
+      puts 'Item cannot be archived'
+    end
+  end
+end
```
Comment (muneebulrehman) [inline item.rb]: Every item should have info about the `Genre`, `Author`, and `Label`. However, in your case, you have selected a generic class for each item. Eg., the `Book` class has info about `Label` and `Genre` but doesn't hold info about `Author`. Kindly fix this issue for all the Item classes.
I have recorded [this video](https://www.loom.com/share/f47a76f418374873bc5dbe41fecf1d17?sid=50059ae5-df2e-4456-acf1-d33542e0bdc7) where i explained in detail about this issue.

File: Method/book.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+class Book < Item
+  attr_accessor :publisher, :cover_state, :title, :color, :labels, :genre
+
+  def initialize(publish_date, title, publisher, color, cover_state)
+    super(publish_date)
+    @title = title
+    @publisher = publisher
+    @color = color
+    @cover_state = cover_state
```
Comment (muneebulrehman) [inline Method/book.rb]: `cover_state` should be a string, however, it is being saved as a boolean. Kindly fix this issue. 
image
image

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+  def to_json(param)
+    {
+      'id' => param.id,
+      'multiplayer' => param.multiplayer,
+      'last_played_at' => param.last_played_at,
+      'publish_date' => param.publish_date,
+      'archived' => param.archived
+    }
+  end
```
Comment (muneebulrehman) [inline app.rb:112]: When listing a `Game` or `music album` or `book`, it should provide all the information about that particular item ie., info about the `Author`, `Genre`, and `Label`. However, in your case, it doesn't provide all the information. Kindly fix this issue. 
Have a look at the below example when i list the books.
image
Even though there is a lot more information about the book in the storage which in itself is not complete, however, is not being displayed in the console. 
image

**Review by muneebulrehman — CHANGES_REQUESTED** [review]
Hi team, 

Good job so far! I see you put a lot of effort into this project. However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- No linter errors
- Descriptive PR
- Good docs
- Great job with the project so far. 👍

_Note_: Kindly ignore my statement on the `source` class in the video. It is only a requirement for a group of 4. However, all the other requirements should be met and are not optional.

## Required Changes ♻️
_Check the comments under the review._

One of the project requirements was to create a video with a max length of 5 minutes. Kindly re-record the video making sure the length of the video is not more than 5 minutes.
Also, make sure you talk about a piece of code that you are proud of.
image


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 1 above)
Comment (Metaverse-Nabeel) [reply README.md]: Hello @muneebulrehman 

Thank you so much for your kind feedback. We will make the required changes and shall again ask for a review.

Regards

File: app.rb (hunk 4 above)
Comment (smunoz1988) [reply app.rb:112]: @muneebulrehman / @Metaverse-Nabeel @Booze33 

Thanks for the review. We will make the changes and aks for another review!

File: item.rb (hunk 2 above)
Comment (Metaverse-Nabeel) [reply item.rb]: Hello @muneebulrehman 

I hope you are doing well. I am thankful that you have done an in-depth review of our project. But we believe there are a few things that need your attention. 

- For team member 1, `Book` should have a book class and label class, not any other class like `genre` or `author`. I have attached a screenshot for your reference. 

image

- For team member 2, `Music Album` should have a music_album class and genre class, not any other class like `lable` or `author`. I have attached a screenshot for your reference. 

image

- For team member 3, `Game` should have a game class and author class, not any other class like `genre` or `lable`. I have attached a screenshot for your reference. 

image

File: Method/book.rb (hunk 3 above)
Comment (Metaverse-Nabeel) [reply Method/book.rb]: Hello @muneebulrehman 

Thank you for your feedback. The changes have been made.

Regards

File: app.rb (hunk 4 above)
Comment (Metaverse-Nabeel) [reply app.rb:112]: Hello @muneebulrehman 

I have made the changes as per the reviews.

File: item.rb (hunk 2 above)
Comment (cynthiainga) [inline item.rb]: - [x] Good job implementing all the requested changes, but what the previous reviewer trying to say is the associations (1-to-many relationships) should be implemented,  `Item` class should have one `Genre`, `Author`, and `Label`.
It should be something like 👇🏾 👇🏾 

```ruby
def genre=(genre)
      @genre = genre
      genre.items.push(self) unless genre.items.include?(self)
    end
```
And that for `author` and `label` as well 😁

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Highlights

- [ ] All tests are passing ✔️ 
- [ ] No linters errors ✔️ 
- [ ] Good commit history ✔️ 

 ## Required Changes ♻️

- [ ] Please kindly check the comment under the review 👇🏾 👇🏾 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: item.rb (hunk 2 above)
Comment (Metaverse-Nabeel) [reply item.rb]: Hello @cynthiainga 

I hope you are doing well. Thank you so much for clarifying the important requirement. We have made the requested changes and now will submit them for review again.

Regards

**Review by OLIPLICHE — APPROVED** [review]
Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Highlights
- Fixed the issues pointed out by the previous reviewer👏
- Class Item has one genre, Author and Label👏
- The App is created successfully 👏 👏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**